### ISSUE TITLE
Let face assemblers describe that they need strain rate and viscosity

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#1388]
@jaustermann ran into this problem with face assemblers: In `assembly.cc`, we have this code in `local_assemble_stokes_system()`:
```
    // then also work on possible face terms. if necessary, initialize
    // the material model data on faces
    for (unsigned int face_no=0; face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
      if (cell->at_boundary(face_no))
        {
          scratch.face_finite_element_values.reinit (cell, face_no);

          if (assemblers->stokes_system_assembler_on_boundary_face_properties.need_face_material_model_data)
            {
              compute_material_model_input_values (current_linearization_point,
                                                   scratch.face_finite_element_values,
                                                   cell,
                                                   rebuild_stokes_matrix,    // *********************************
                                                   scratch.face_material_model_inputs);
              create_additional_material_model_outputs(scratch.face_material_model_outputs);

              material_model->evaluate(scratch.face_material_model_inputs,
                                       scratch.face_material_model_outputs);
```

In other words, we only evaluate the strain rate and (and consequently) viscosity if the face assemblers are used to (re-)build the matrix. But in her case, she just wants to assemble a rhs term that happens to require the viscosity, without rebuilding the matrix.

We ought to find a way to let the face assemblers describe what they actually need.

**Comment by tjhei** [comment]
> We ought to find a way to let the face assemblers describe what they actually need.

My vote would be to move away from only using signals and use the base class to query various things.

**Comment by gassmoeller** [comment]
I agree. When looking into #1385 I also noticed our current system is not optimal. `set_assemblers` is the only place that currently knows all information to set up properties for all assemblers, and it will grow over time with more and more options. This should be done inside the individual assemblers, similar to the `model_dependence` member of the material model interface.

**Comment by gassmoeller** [comment]
Closed by #1385.
